sdks/python/apache_beam/typehints/decorators.py

```diff
@@ -217,7 +220,10 @@ def get_signature(func):
   return signature
 
 
-class IOTypeHints(object):
+class IOTypeHints(NamedTuple('IOTypeHints', [
+    ('input_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
+    ('output_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
+    ('origin', List[str])])):
```

Comment:
My guess is that this is going to trip up mypy, but we can defer that to a future mypy PR if you'd like.  This is the pattern that I use for these cases:

```python
_IOTypeHints = NamedTuple('_IOTypeHints', [
    ('input_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
    ('output_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
    ('origin', List[str])
])

class IOTypeHints(_IOTypeHints):
```

Reply:
The tradeoff is that this produces a sub-par repr for `IOTypeHints`:

```python
In [2]: from typing import *

In [3]: _IOTypeHints = NamedTuple('_IOTypeHints', [
   ...:     ('input_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
   ...:     ('output_types', Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]),
   ...:     ('origin', List[str])
   ...: ])

In [4]: class IOTypeHints(_IOTypeHints):pass

In [5]: x = IOTypeHints(None, None, ['foo'])

In [6]: x
Out[6]: _IOTypeHints(input_types=None, output_types=None, origin=['foo'])
```

Doing a quick investigation to see if there's a hack fix for this.

Reply:
Something like this:

```python
    def __repr__(self):
      repr_fmt = '{typename}({repr_fmt})'.format(
            typename=type(self).__name__, 
            repr_fmt=', '.join('{name}=%r'.format(name=name) for name in self._fields)
      )
      return repr_fmt % self
```

Reply:
Trip up how?

I see this error in lint (should be fixed in next commit):
```
apache_beam/typehints/decorators.py:353: error: Value of type "Optional[Tuple[Tuple[Any, ...], Dict[str, Any]]]" is not indexable  [index]
```
This is the line:
```
output_type = self.output_types[0][0]
```
So it seems that mypy is correctly parsing these type hints.

Reply:
I assumed that mypy would need the intermediate class _or_ would want that class's name to match the name provided to `NamedTuple()`.  For example, both of these are invalid:

```python
class Foo(Generic[TypeVar('T')]):  # TypeVar needs to be defined above
    pass

T = TypeVar('T_with_suffix')  # TypeVar name needs to match variable name
class Bar(Generic[T]):
    pass
```

I just did my own isolated test of this and you're right, mypy has no issue with it, so it appears there's nothing special about the name given to `NamedTuple`, as far as mypy is concerned.

Thanks for taking the time to test that.  I was trying to give some quick feedback without getting too drawn in, but once I _did_ get drawn in, I should have gone back and checked my original assumption!